### ISSUE TITLE
Cannot link on Windows (mingw) because of time_now

### ISSUE THREAD

**Issue by jfthuong** [janestreet/ppx_inline_test#23]
Version 0.14.0 is using `time_now`.

It cannot link on Windows when running inline tests (tried with dune)

![image](https://user-images.githubusercontent.com/1401809/84558653-0aa45500-ad67-11ea-9d3a-7cf43f378f27.png)

Error returned is the following:
```
(cd _build/default && C:\OCaml64\home\jthuong\.opam\4.10.0+mingw64c\bin\ocamlopt.opt.exe -w -24 -g -o test/.test_inline.inline-tests/run.exe -linkall C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\stdlib-shims\stdlib_shims.cmxa C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\fmt\fmt.cmxa C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\astring\astring.cmxa C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\cmdliner\cmdliner.cmxa C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\fmt\fmt_cli.cmxa C:/OCaml64/home/jthuong/.opam/4.10.0+mingw64c/lib/ocaml\unix.cmxa -I C:/OCaml64/home/jthuong/.opam/4.10.0+mingw64c/lib/ocaml C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\fmt\fmt_tty.cmxa C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\uuidm\uuidm.cmxa C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\re\re.cmxa C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\alcotest\alcotest.cmxa C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\base\base_internalhash_types/base_internalhash_types.cmxa -I C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\base\base_internalhash_types C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\base\caml/caml.cmxa C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\sexplib0\sexplib0.cmxa C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\base\shadow_stdlib/shadow_stdlib.cmxa C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\base\base.cmxa -I C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\base C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\ppx_inline_test\config/inline_test_config.cmxa C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\jane-street-headers\jane_street_headers.cmxa C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\ppx_sexp_conv\runtime-lib/ppx_sexp_conv_lib.cmxa C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\ppx_compare\runtime-lib/ppx_compare_lib.cmxa C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\ppx_enumerate\runtime-lib/ppx_enumerate_lib.cmxa C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\ppx_hash\runtime-lib/ppx_hash_lib.cmxa C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\time_now\time_now.cmxa -I C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\time_now C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\ppx_inline_test\runtime-lib/ppx_inline_test_lib.cmxa test/test_inline.cmxa C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\ppx_inline_test\runner/lib/ppx_inline_test_runner_lib.cmxa -I C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\ppx_inline_test\runner/lib test/.test_inline.inline-tests/.test_inline.inline-tests.eobjs/native/dune__exe__Run.cmx)
** Cannot resolve symbols for C:\\OCaml64\\home\\jthuong\\.opam\\4.10.0+mingw64c\lib\time_now\libtime_now_stubs.a(time_now_stubs.o/
):
 clock_gettime
File "caml_startup", line 1:
Error: Error during linking
```

I am using
* `dune` version 2.5.1
* `time_now` version 0.14.0
* compiler `ocaml-variants.4.10.0+mingw64c`

My mini project is the following:
[dune_test.zip](https://github.com/janestreet/ppx_inline_test/files/4773917/dune_test.zip)

Thank you @jeremiedimino for pointing me in the right direction.

**Comment by jfthuong** [comment]
Currently, the workaround is to pin to version `v0.13.1`.

**Comment by jfthuong** [comment]
... on the other hand, version **0.13.1** fails with `ppx_expect as follows`:

Inline Test (test is in line 1 of *test_inline.ml*)
```ocaml
let%expect_test _ =
  print_endline "Hello, world!";
  [%expect{| |}]

```

Error message
```
File "test/test_inline.ml", line 1, characters 0-71 threw End_of_file.
  Raised at file "stdlib.ml", line 422, characters 9-26
  Called from file "stdlib.ml", line 433, characters 2-25
  Called from file "collector/expect_test_collector.ml", line 110, characters 14-40
  Called from file "collector/expect_test_collector.ml", line 141, characters 26-60
  Called from file "list.ml", line 121, characters 24-34
  Called from file "collector/expect_test_collector.ml", line 137, characters 14-256
  Called from file "collector/expect_test_collector.ml", line 20, characters 8-12
  Re-raised at file "collector/expect_test_collector.ml", line 26, characters 4-11
  Called from file "collector/expect_test_collector.ml", line 20, characters 8-12
  Re-raised at file "collector/expect_test_collector.ml", line 26, characters 4-11
  Called from file "collector/expect_test_collector.ml", line 239, characters 51-76
  Called from file "collector/expect_test_collector.ml", line 237, characters 12-602
  Called from file "collector/expect_test_collector.ml", line 301, characters 11-90
  Called from file "runtime-lib/runtime.ml", line 343, characters 8-12
  Re-raised at file "runtime-lib/runtime.ml", line 346, characters 6-13
  Called from file "runtime-lib/runtime.ml", line 359, characters 15-52
  Called from file "runtime-lib/runtime.ml", line 446, characters 52-83
```

**Comment by jeremiedimino** [comment]
I had a quick look at this. @jfthuong I believe that if you add `core_kernel` as a dependency of your test the error with v0.14 will go away.

For JS developers: it seems to me that the `nanoseconds_since_unix_epoch` functions were moved from `core_kernel` to a new standalone `time_now` library recently. However, the `(:include rt-flags)` was left in the jbuild file of `core_kernel`. The `rt-flags` is created by `lib/config_h/discover/discover.ml`, and contains:

- nothing on systems that don't require `-lrt` for `clock_gettime`
- `-lrt` on systems that require `-lrt` for `clock_gettime`

so if the stubs for `clock_gettime` now live in `time_now`, the `(:include rt-flags)` should live there as well.

**Comment by jfthuong** [comment]
Thanks for taking the time to check.

Do you mean something like this
```dune
(library
 (name test_inline)
 (libraries core_kernel)
 (inline_tests)
 (preprocess (pps ppx_inline_test)))
```

If so, I still have the same linking error (I also tried to add `core_kernel` dependency in `inline_tests` stanza).

**Comment by jeremiedimino** [comment]
Yeah, that's what I meant. Hmm, that's odd then :thinking:

**Comment by sliquister** [comment]
Can you try to build a program on windows that runs `Time_now.nanoseconds_since_unix_epoch ()`? If that fails, then this let's move this issue to the time_now library instead.

Otherwise, it seems more complicated

**Comment by sliquister** [comment]
I'm going to fix the problem that @jeremiedimino noticed with the -lrt flag being in core_kernel instead of time_now, and hope that it fixes the issue (maybe the workaround of adding core_kernel to the list of libs didn't help due to flag order), and close the issue unless someone can check whether the issue is still there or not.

**Comment by aalekseyev** [comment]
This change landed in 4ad759e0d454fba07007901f5221214cb4cd8cd5.
I'm closing the issue, assuming it's fixed, but feel free to reopen especially since as @sliquister says it's not tested on Windows.

**Comment by aalekseyev** [comment]
Actually the change is not in ppx_inline_test repo, so you won't find it in the commit I mentioned above. The change should be there in the relevant repos, though.
